### PR TITLE
migrate LSR scripts dev

### PR DESCRIPTION
**Pull request by geforcelive**

### CODE REVIEW

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 1]
```diff
@@ -53,18 +55,11 @@ function main() {
     delete_operator "ibm-cert-manager-operator" "$OPERATOR_NS"
     
     if [[ $ENABLE_LICENSING -eq 1 ]]; then
-
-        is_exists=$("$OC" get deployments ibm-licensing-operator -n "$OPERATOR_NS")
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh]: Looks like the code to migrate License Service is now removed, so are we regressing here the License Service upgrade, by adding Reported migrate? Pls clarify - thx.

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 2]
```diff
@@ -81,6 +76,88 @@ function main() {
     success "Migration is completed for Cloud Pak 3.0 Foundational singleton services."
 }
 
+function migrate_license_service_reporter(){
+    title "LSR migration from ibm-cmmon-services to ${LSR_NAMESPACE}"
+
+    local lsr_instances=$("$OC" get IBMLicenseServiceReporter instance -n ibm-common-services --no-headers | wc -l)
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:93]: We shall not assume it's alays `ibm-common-services` - i.e. sometimes the customer / Pak could have configured the namespace into the custom namespace name

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 3]
```diff
@@ -81,6 +76,88 @@ function main() {
...
+    if [[ lsr_instances -eq 0 ]]; then
+        info "No LSR for migration found in ibm-common-services namespace"
+        return 0
+    fi
+    # Prepare LSR PV/PVC which was decoupled in isolate.sh
+    # delete old LSR CR - PV will stay as during isolate.sh the policy was set to Retain
+    ${OC} delete IBMLicenseServiceReporter instance -n ibm-common-services
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:100]: Does it make sense to change the order, i.e. check the PV was labelled, before we remove `IBMLicenseServiceReporter` CR? If for some reason adopter hasn't picked the right order, the data is lost without any recovery.

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 4]
```diff
@@ -81,6 +76,88 @@ function main() {
...
+        metadata:
+            name: license-service-reporter-pvc
+            namespace: ${LSR_NAMESPACE}
+        spec:
+            accessModes:
+            - ReadWriteOnce
+            resources:
+                requests:
+                    storage: 1Gi
+            storageClassName: "${LSR_STORAGE_CLASS}"
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:151]: Does empty storageClassName works on ROKS? Or it should not be added at all?

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 5]
```diff
@@ -81,6 +76,88 @@ function main() {
...
+            volumeMode: Filesystem
+            volumeName: ${LSR_PV_NAME}
+EOF
+
+        ${OC} create -f ${TEMP_LSR_PVC_FILE}
+        # checking status of PVC - in case it cannot be boud, the claimRef needs to be set to null
+        status=$("${OC}" get pvc license-service-reporter-pvc -n $LSR_NAMESPACE --no-headers | awk '{print $2}')
+        while [[ "$status" != "Bound" ]]
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:159]: Do we really need to wait for the PVC binding, given we don't depend on it at all here? For mongo, we had to do the wait time, since we load data into it, but here we don't really need to do so, right?

**Review by pgodowski — CHANGES_REQUESTED** [review]

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 6]
```diff
@@ -92,14 +180,30 @@ function restore_ibmlicensing() {
...
-        with(.; del(.metadata.creationTimestamp) |
-        del(.metadata.managedFields) |
-        del(.metadata.resourceVersion) |
-        del(.metadata.uid) |
-        del(.status)
-        )
-    ' | sed -e 's/^/    /g'`
+    # for LS + LSR installed in the same cluster, the sender needs to be removed as it is not valid any more after migration
+    local reporterURL=$("${OC}" get IBMLicensing instance  --ignore-not-found -o=jsonpath='{.spec.sender.reporterURL}')
+    if [[ $reporterURL == "https://ibm-license-service-reporter:8080" ]]; then
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:203]: Is this the only variant indicating the local Reporter?

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 7]
```diff
@@ -168,6 +279,9 @@ function print_usage() {
     echo "   --yq string                                    File path to yq CLI. Default uses yq in your PATH"
     echo "   --operator-namespace string                    Required. Namespace to migrate Foundational services operator"
     echo "   --enable-licensing                             Set this flag to migrate ibm-licensing-operator"
+    echo "   --enable-license-service-reporter              Set this flag to install ibm-license-service-reporter-operator"
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:305]: Imrpove the description to add human name of Reporter

File: cp3pt0-deployment/common/utils.sh [hunk 8]
```diff
@@ -333,7 +333,7 @@ function wait_for_nss_patch() {
     fi
     
     # wait for deployment to be ready
-    local deployment_name=$(${OC} get deployment -n ${namespace} -l operators.coreos.com/${package_name}.${namespace}='' --no-headers | awk '{print $1}')
+    local deployment_name="ibm-common-service-operator"
```
Comment (pgodowski) [inline cp3pt0-deployment/common/utils.sh]: Why are we changing this, as it does not seem related to Reporter?

File: cp3pt0-deployment/setup_singleton.sh [hunk 9]
```diff
@@ -66,14 +70,19 @@ function main() {
 
     if [ $MIGRATE_SINGLETON -eq 1 ]; then
         if [ $ENABLE_LICENSING -eq 1 ]; then
-            ${BASE_DIR}/common/migrate_singleton.sh "--operator-namespace" "$OPERATOR_NS" --control-namespace "$CONTROL_NS" "--enable-licensing" --licensing-namespace "$LICENSING_NAMESPACE"
+            if [ $ENABLE_LICENSE_SERVICE_REPORTER -eq 1 ]; then
+                ${BASE_DIR}/common/migrate_singleton.sh "--operator-namespace" "$OPERATOR_NS" "--control-namespace" "$CONTROL_NS" "--enable-licensing" "--licensing-namespace" "$LICENSING_NAMESPACE" "--enable-license-service-reporter" "--lsr-namespace" "$LSR_NAMESPACE" "-v" "$DEBUG"
+            else
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_singleton.sh:75]: The earlier version was not passing the verbose flag - we are sure it works for both LS and LSR?

File: cp3pt0-deployment/setup_singleton.sh [hunk 10]
```diff
@@ -162,12 +182,15 @@ function print_usage() {
     echo "   --oc string                                    Optional. File path to oc CLI. Default uses oc in your PATH"
     echo "   --yq string                                    Optional. File path to yq CLI. Default uses yq in your PATH"
     echo "   --operator-namespace string                    Optional. Namespace to migrate Cloud Pak 2 Foundational services"
-    echo "   --enable-licensing                             Optional. Set this flag to install ibm-licensing-operator"
+    echo "   -ls, --enable-licensing                        Optional. Set this flag to install ibm-licensing-operator"
+    echo "   -lsNs, --licensing-namespace string            Optional. Set custom namespace for ibm-licensing-operator. Default is ibm-licensing"
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_singleton.sh:190]: We are breaking the interface here: we change the parameter `-licensingNs` into `--licensing-namespace` which sounds dangerous

File: cp3pt0-deployment/setup_singleton.sh [hunk 11]
```diff
@@ -220,7 +243,7 @@ function is_migrate_licensing() {
     get_and_validate_arguments
     if [ ! -z "$CONTROL_NS" ]; then
         if [[ "$CUSTOMIZED_LICENSING_NAMESPACE" -eq 1 ]] && [[ "$CONTROL_NS" != "$LICENSING_NAMESPACE" ]]; then
-            error "Licensing Migration could only be done in $CONTROL_NS, please do not set parameter '-licensingNs $LICENSING_NAMESPACE'"
+            error "Licensing Migration could only be done in $CONTROL_NS, please do not set parameter '-lsNs $LICENSING_NAMESPACE'"
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_singleton.sh:254]: see above comment - it seems we're changing the parameters name, which looks like a breaking  interface change

File: cp3pt0-deployment/setup_tenant.sh [hunk 12]
```diff
@@ -253,7 +253,7 @@ function default_arguments() {
     # check if CommonService CRD exists in cluster
     local is_CS_CRD_exist=$((${OC} get commonservice -n ${OPERATOR_NS} --ignore-not-found > /dev/null && echo exists) || echo fail)
     if [[ "$is_CS_CRD_exist" == "exists" ]]; then
-        result=$("${OC}" get commonservice common-service -n ${OPERATOR_NS} -o yaml --ignore-not-found)
+        local result=$("${OC}" get commonservice common-service -n ${OPERATOR_NS} -o yaml --ignore-not-found)
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_tenant.sh]: why we are changing the code not relevant to LSR?

File: cp3pt0-deployment/setup_tenant.sh [hunk 13]
```diff
@@ -270,6 +270,13 @@ function default_arguments() {
         else
             info "CommonService CRD exists but main CommonService CR does not exist, skipping defaulting from original CommonService CR\n"
         fi
+
+        # if CommonService CRD exists and subscription of common-service-operator exists, simple topology will be accepted
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_tenant.sh]: Are we sure we want to add the code not relevant to LSR?

File: cp3pt0-deployment/setup_tenant.sh [hunk 14]
```diff
@@ -541,7 +548,6 @@ function install_cs_operator() {
         fi
         wait_for_operator "$OPERATOR_NS" "ibm-common-service-operator"
         accept_license "commonservice" "$OPERATOR_NS" "common-service"
-        wait_for_cs_webhook "$OPERATOR_NS" "ibm-common-service-webhook"
```
Comment (pgodowski) [inline cp3pt0-deployment/setup_tenant.sh]: why this is removed?

**Review by pgodowski — CHANGES_REQUESTED** [review]

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 1 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh]: restored

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 2 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh]: changed to ${OPERATOR_NS}

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 3 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh]: moved check for LSR PV to earlier stage

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 5 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh:160]: From our tests it seems we need to wait. @martabereta could you comment on details?

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 4 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh:152]: We was about to check it. @martabereta could you comment on details?

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 6 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh:204]: It should be the only one. @martabereta please comment

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 7 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/migrate_singleton.sh]: improved

File: cp3pt0-deployment/common/utils.sh (hunk 8 above)
Comment (geforcelive) [reply cp3pt0-deployment/common/utils.sh]: it was merge from origin repo

File: cp3pt0-deployment/setup_singleton.sh (hunk 9 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_singleton.sh:75]: @martabereta please comment

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 5 above)
Comment (martabereta) [reply cp3pt0-deployment/common/migrate_singleton.sh:160]: Here are the logs from the tests - it looks that the loop was in use:
```
# Checking whether Namespace lsr-ns exist...
[INFO] Creating namespace lsr-ns
namespace/lsr-ns created
[INFO] Waiting for project lsr-ns to be created
[✔] Project lsr-ns is created

persistentvolumeclaim/license-service-reporter-pvc created
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched
[INFO] Waiting for pvc license-service-reporter-pvc to bind
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched
[INFO] Waiting for pvc license-service-reporter-pvc to bind
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched
[INFO] Waiting for pvc license-service-reporter-pvc to bind
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched
[INFO] Waiting for pvc license-service-reporter-pvc to bind
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched
[INFO] Waiting for pvc license-service-reporter-pvc to bind
persistentvolume/pvc-f5b2f55f-0931-41d3-b60c-c2ae05649d66 patched (no change)
[INFO] Waiting for pvc license-service-reporter-pvc to bind
[INFO] Waiting for pvc license-service-reporter-pvc to bind
configmap/ibmlicensing-instance-bak created
ibmlicensing.operator.ibm.com "instance" deleted
# Deleting ibm-licensing-operator in namesapce ibm-common-services...
[1] Removing the subscription of ibm-licensing-operator in namesapce ibm-common-services ...
[2] Removing the csv of ibm-licensing-operator in namesapce ibm-common-services ...```

File: cp3pt0-deployment/setup_singleton.sh (hunk 10 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_singleton.sh]: reverted to -licensingNs

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 6 above)
Comment (martabereta) [reply cp3pt0-deployment/common/migrate_singleton.sh:204]: We've discussed it with @szymon-kowalczyk - this is the only option we are aware of

File: cp3pt0-deployment/setup_singleton.sh (hunk 11 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_singleton.sh]: reverted to -licensingNs

File: cp3pt0-deployment/setup_tenant.sh (hunk 12 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_tenant.sh]: it is merge from origin repo

File: cp3pt0-deployment/setup_singleton.sh (hunk 10 above)
Comment (martabereta) [reply cp3pt0-deployment/setup_singleton.sh]: fair point

File: cp3pt0-deployment/setup_tenant.sh (hunk 13 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_tenant.sh]: it is merge from origin repo

File: cp3pt0-deployment/setup_tenant.sh (hunk 14 above)
Comment (geforcelive) [reply cp3pt0-deployment/setup_tenant.sh]: it is merge from origin repo

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 5 above)
Comment (pgodowski) [reply cp3pt0-deployment/common/migrate_singleton.sh:160]: I am not asking what the code does. I am asking why do we need to wait at all, if in our script, we don't make any use of the created PVC? It should be up to Reporter operator to deal with that, right?

**Review by pgodowski — CHANGES_REQUESTED** [review]

**Comment by pgodowski** [comment]
It looks good to me from the Licensing side perspective. @bitscuit can u pls review ?

File: cp3pt0-deployment/common/migrate_singleton.sh (hunk 5 above)
Comment (martabereta) [reply cp3pt0-deployment/common/migrate_singleton.sh:160]: The loop contains not only checking if PVC was bound but in case it is not it bound, it set `claimRef` on PV to null

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 15]
```diff
@@ -81,6 +87,106 @@ function main() {
...
+            deprecated_region_label='failure-domain.beta.kubernetes.io/region'
+            not_deprecated_region_label='topology.kubernetes.io/region'
+            deprecated_zone_label='failure-domain.beta.kubernetes.io/zone'
+            not_deprecated_zone_label='topology.kubernetes.io/zone'
+
+            region=$(oc get pv -l license-service-reporter-pv=true -o=jsonpath=$deprecated_region)
+            zone=$(oc get pv -l license-service-reporter-pv=true -o=jsonpath=$deprecated_zone)
+
+            if [[ $region != "" ]]; then
+                debug1 "Replacing depracated PV labels"
```
Comment (pgodowski) [inline cp3pt0-deployment/common/migrate_singleton.sh:149]:
```suggestion
                debug1 "Replacing deprecated PV labels"
```

**Review by pgodowski — APPROVED** [review]
